src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificateAssetDownloader.cs

```diff
@@ -75,99 +73,37 @@ internal static SafeX509CrlHandle DownloadCrl(string uri, ref TimeSpan remaining
 
         private static byte[] DownloadAsset(string uri, ref TimeSpan remainingDownloadTime)
         {
-            if (remainingDownloadTime <= TimeSpan.Zero)
-            {
-                return null;
-            }
-
-            List<byte[]> dataPieces = new List<byte[]>();
-
-            using (Interop.Http.SafeCurlHandle curlHandle = Interop.Http.EasyCreate())
+            if (remainingDownloadTime > TimeSpan.Zero)
             {
-                GCHandle gcHandle = GCHandle.Alloc(dataPieces);
-                Interop.Http.SafeCallbackHandle callbackHandle = new Interop.Http.SafeCallbackHandle();
-
+                Stopwatch stopwatch = Stopwatch.StartNew();
+                object httpClient = null;
                 try
                 {
-                    Interop.Http.EasySetOptionString(curlHandle, Interop.Http.CURLoption.CURLOPT_URL, uri);
-                    Interop.Http.EasySetOptionLong(curlHandle, Interop.Http.CURLoption.CURLOPT_FOLLOWLOCATION, 1L);
-
-                    IntPtr dataHandlePtr = GCHandle.ToIntPtr(gcHandle);
-                    Interop.Http.RegisterReadWriteCallback(
-                        curlHandle,
-                        Interop.Http.ReadWriteFunction.Write,
-                        s_writeCallback,
-                        dataHandlePtr,
-                        ref callbackHandle);
-
-                    Stopwatch stopwatch = Stopwatch.StartNew();
-                    Interop.Http.CURLcode res = Interop.Http.EasyPerform(curlHandle);
-                    stopwatch.Stop();
-
-                    // TimeSpan.Zero isn't a worrisome value on the subtraction, it only
-                    // means "no limit" on the original input.
-                    remainingDownloadTime -= stopwatch.Elapsed;
-
-                    if (res != Interop.Http.CURLcode.CURLE_OK)
+                    // Use reflection to access System.Net.Http:
+                    // Since System.Net.Http.dll explicitly depends on System.Security.Cryptography.X509Certificates.dll,
+                    // the latter can't in turn have an explicit dependency on the former.
+                    Type httpClientType = Type.GetType("System.Net.Http.HttpClient, System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: false);
+                    if (httpClientType != null)
                     {
-                        return null;
+                        MethodInfo getByteArrayAsync = httpClientType.GetMethod("GetByteArrayAsync", new Type[] { typeof(string) });
```

Comment:
Would promoting this to a `Lazy<MethodInfo>` make sense?

If we don't have any scenarios where we expect this reflection to fail, we should perhaps make it be more throwy.

Does this require any rd.xml changes for aot?

Reply:
> Does this require any rd.xml changes for aot?

HttpClient.GetByteArrayAsync is a public API, so... no?

> Would promoting this to a Lazy<MethodInfo> make sense?

I don't have a sense for how common this operation is.  Seemed like it would be pretty rare, and so caching all of this wouldn't be necessary.  If it's more hot-path-y, I'd want to do more here to alleviate costs, like keep the HttpClient instance around to be reused.

Reply:
AFAIK, aot runs trimmed, so if you built https://github.com/bartonjs/SampleProjects/blob/master/PrintChain/PrintChain.cs aot it would not include HttpClient at all.  Then the reflection would fail.

Assuming, of course, we have aot for Linux or macOS (which I don't think we do).

Reply:
> AFAIK, aot runs trimmed, so if you built https://github.com/bartonjs/SampleProjects/blob/master/PrintChain/PrintChain.cs aot it would not include HttpClient at all. Then the reflection would fail. Assuming, of course, we have aot for Linux or macOS (which I don't think we do).

Ah, I see.  We use similar reflection in other places on Linux and don't include rd.xml stuff, like using System.Diagnostics.Process from Environment.WorkingSet.  Seems like if/when that becomes an issue we can address it for all of them.  @jkotas, is this something we should worry about now?

Reply:
This is the general problem for IL trimming. I do not think this makes it worse in any way.

The change to use reflection looks fine to me.